system/jlib/jsocket.cpp

```diff
@@ -4646,7 +4646,12 @@ class CSocketEpollThread: public CSocketBaseThread
                             }
 # endif
                             if (epevents[j].data.fd >= 0) {
-                                assertex(epfdtbl[epevents[j].data.fd] >= 0);
+                                // assertex(epfdtbl[epevents[j].data.fd] >= 0);
+                                if (epfdtbl[epevents[j].data.fd] < 0)
```

Comment:
Still very unsure how this can happen..

updateEpollVars/updateItems(), is only called on this thread and proceeds the epoll_wait.
It is the only thing that can have updated epfdtbl (set an index to -1) and in doing so it will have called epoll_op to update epoll.

i.e. it looks thread safe and epoll should be correctly updated before the epoll_wait and guarantee not to return a epevents[j].data.fd that is out of sync with epfdtbl[]...

If it can be out of sync, then could it also miss events? i.e. recently added file descriptors, also updated by updateItems(), not generating events. Perhaps in that case, it would simply catch them the next time epoll_wait is called.

It's a bit mysterious.